eng/testing/tests.mobile.targets

```diff
@@ -109,6 +107,6 @@
 
   <Target Name="PublishTestAsSelfContained"
```

Comment:
Wouldn't it make more sense to invoke the publish target instead of sequencing this target after build and making it dependent on the `Publish` target?

Reply:
Where would you invoke it? I did it this way as we need to update the runtime pack first and then run the app bundle generation. Also, I needed to hook it up into after build because `Publish` depends on build and to avoid a circular dependency. 

I'm open to ideas though, could you elaborate on what you have in mind?

Reply:
> hook it up into after build because Publish depends on build and to avoid a circular dependency.

The publish target already depends on the Build target: https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L57. Unsure if we are talking about the same thing here?

> Where would you invoke it? I did it this way as we need to update the runtime pack first and then run the app bundle generation.

I was assuming we can just invoke the /t:Publish target (or dotnet publish) on the test projects and sequence the other targets in this file into it.

Reply:
That would make inner dev loop different for mobile/xamarin tests which I wanted to avoid. I wanted to have the same experience, if I do `dotnet build` or `dotnet build /t:test` on a test project and target `xamarin` the infra knows what to do, not me as a developer having to remember that for that I have to do `/t:publish`.

> The publish target already depends on the Build target: https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L57. Unsure if we are talking about the same thing here?

Yeah, that's why I have to hook up after the build target to avoid circular dependency errors.

Reply:
> That would make inner dev loop different for mobile/xamarin tests which I wanted to avoid. I wanted to have the same experience, if I do dotnet build or dotnet build /t:test on a test project and target xamarin the infra knows what to do, not me as a developer having to remember that for that I have to do /t:publish.

I see, makes sense. FWIW aspnetcore is using publish on their test apps for helix submission. I plan to do the same for our tests and get rid of `/p:ArchiveTests`. Sounds like when we do that, we will need to move this piece over as well which I'm fine with.

Reply:
Yeah, whenever we do that it is just a matter of moving the pieces to hook up these mobile app targets together and we could reuse the logic I used to tell the SDK here is my runtime pack for you to use to publish as self contained.

Reply:
As a side not for those plans, when publishing as a self contained, you need to have the runtime pack for that RID, so that won't allow us sharing the build of tests on multiple RID test runs, so something to think about and that we might need to change in our CI/PR builds if we want to strive towards that. 

For Mono legs, I plan on removing the run test legs and just building the tests and running them as part of the same legs.

Reply:
> As a side not for those plans, when publishing as a self contained, you need to have the runtime pack for that RID, so that won't allow us sharing the build of tests on multiple RID test runs

When I was talking about switching to use the Publish target I didn't mean to switch to self-contained as well. I don't think that would apply here, right?

FWIW we will need to publish self-contained at least for code coverage in the future (talking about local inner-loop) as the current model is easy to break (we instrument referenced assemblies from the testhost directory and then undo the instrumentation when the coverage run is finished). In the case of something going wrong or the process being killed, the testhost is broken and you need to rebuild.

Reply:
> When I was talking about switching to use the Publish target I didn't mean to switch to self-contained as well. I don't think that would apply here, right?

Oh right.

> FWIW we will need to publish self-contained at least for code coverage in the future (talking about local inner-loop) as the current model is easy to break (we instrument referenced assemblies from the testhost directory and then undo the instrumentation when the coverage run is finished). In the case of something going wrong or the process being killed, the testhost is broken and you need to rebuild.

Makes sense, well this is giving a way towards that.